File: android/app/src/main/java/com/guardforce/guard/domain/SemVer.kt
package com.guardforce.guard.domain

object SemVer {
    private fun parts(v: String): List<Int> = v.trim().removePrefix("v").split('.', '-', '+').take(3).map { it.toIntOrNull() ?: 0 }

    /** Negative when [a] < [b]. Missing segments count as zero, so "1.2" == "1.2.0". */
    fun compare(a: String, b: String): Int {
        val pa = parts(a); val pb = parts(b)
        for (i in 0 until 3) {
            val x = pa.getOrElse(i) { 0 }; val y = pb.getOrElse(i) { 0 }
            if (x != y) return x.compareTo(y)
        }
        return 0
    }

    fun needsUpdate(current: String, minimum: String?): Boolean = minimum != null && compare(current, minimum) < 0
}

File: guard-app/src/domain/semver.ts
const parts = (v: string) => v.trim().replace(/^v/, "").split(/[.\-+]/).slice(0, 3).map((x) => parseInt(x, 10) || 0);
/** Negative when a < b; "1.2" equals "1.2.0". */
export function compareVersions(a: string, b: string): number {
  const pa = parts(a), pb = parts(b);
  for (let i = 0; i < 3; i++) { const x = pa[i] ?? 0, y = pb[i] ?? 0; if (x !== y) return x < y ? -1 : 1; }
  return 0;
}
export const needsUpdate = (current: string, minimum?: string | null) => !!minimum && compareVersions(current, minimum) < 0;
